File: 1942-the-number-of-the-smallest-unoccupied-chair/1942-the-number-of-the-smallest-unoccupied-chair.py
class Solution:
    def smallestChair(self, times: List[List[int]], targetFriend: int) -> int:
        pq = []
        chairs = []
        
        targetArrival = times[targetFriend][0]
        times.sort(key=lambda x: x[0])
        
        nextChair = 0

        for arrival, leave in times:
            while pq and pq[0][0] <= arrival:
                heapq.heappush(chairs, heapq.heappop(pq)[1])
            
            if chairs:
                seat = heapq.heappop(chairs)
            else:
                seat = nextChair
                nextChair += 1
            
            heapq.heappush(pq, (leave, seat))

            if arrival == targetArrival:
                return seat

File: 1942-the-number-of-the-smallest-unoccupied-chair/1942-the-number-of-the-smallest-unoccupied-chair.java
import java.util.*;

class Solution {
    public int smallestChair(int[][] times, int targetFriend) {
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        PriorityQueue<Integer> chairs = new PriorityQueue<>();
        
        int targetArrival = times[targetFriend][0];
        Arrays.sort(times, Comparator.comparingInt(a -> a[0]));

        int nextChair = 0, seat = 0;

        for (int[] time : times) {
            int arrival = time[0], leave = time[1];

            // Free up chairs of friends who have already left
            while (!pq.isEmpty() && pq.peek()[0] <= arrival) {
                chairs.offer(pq.poll()[1]);
            }

            // Assign the smallest available chair
            if (!chairs.isEmpty()) seat = chairs.poll();
            else seat = nextChair++;

            pq.offer(new int[] {leave, seat});

            // If this is the target friend, return their seat
            if (arrival == targetArrival) break;
        }

        return seat;
    }
}
